nodejs/device/src/index.ts

```diff
@@ -379,85 +435,153 @@ class DeviceApp {
-          serviceAckType: "telemetry",
-          sessionMetrics: this.sessionMetrics,
-        },
-      })
-    );
-    logger.info(`Sending message with serviceAckId: ${serviceAckId}`);
-    this.client
-      .sendEvent(message)
-      .catch(() => {
-        logger.error(
-          `Error sending message with serviceAckId: ${serviceAckId}`
+  private sendTelemetry(): [Promise<void>, () => void] {
+    logger.info("Starting telemetry sending.");
+    let cancel: () => void;
+    const promise = new Promise<void>((_, reject) => {
+      let cleanedUp = false;
+      let cleanupAndReject = (err: Error) => {
+        cleanedUp = true;
+        clearInterval(interval);
+        reject(err);
+      };
+      cancel = () => {
+        if (cleanedUp) {
+          return;
+        }
+        logger.warn("Telemetry sending canceled.");
+        cleanupAndReject(new Canceled());
+      };
+      const interval = setInterval(() => {
+        const serviceAckId = uuidv4();
+        const now = new Date();
+        this.sessionMetrics.lastUpdateTimeUtc = now.toISOString();
+        this.sessionMetrics.runTime = msToDurationString(
+          now.getTime() - this.runStart.getTime()
         );
-      })
-      .finally(() => {
-        this.serviceAckWaitList.add(serviceAckId);
-        if (
-          this.serviceAckWaitList.size > settings.sendMessageAllowedFailureCount
-        ) {
-          this.stopDeviceApp(
-            new Error(
-              `The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${settings.sendMessageAllowedFailureCount}.`
-            )
-          );
+        const message = new Message(
+          JSON.stringify({
+            thief: {
+              cmd: "serviceAckRequest",
+              serviceInstanceId: this.serviceInstanceId,
+              runId: runId,
+              serviceAckId: serviceAckId,
+              serviceAckType: "telemetry",
+              sessionMetrics: this.sessionMetrics,
+            },
+          })
+        );
+        logger.info(`Sending message with serviceAckId: ${serviceAckId}`);
+        this.client
+          .sendEvent(message)
+          .catch(() => {
+            logger.error(
+              `Error sending message with serviceAckId: ${serviceAckId}`
+            );
+          })
+          .finally(() => {
+            this.serviceAckWaitList.add(serviceAckId);
+            if (
+              this.serviceAckWaitList.size >
+              settings.sendMessageAllowedFailureCount
+            ) {
+              cleanupAndReject(
+                new Error(
+                  `The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${settings.sendMessageAllowedFailureCount}.`
+                )
+              );
+            }
+          });
+      }, 1000 * settings.sendMessageOperationsPerSecond);
+    });
+    return [promise, cancel];
+  }
+
+  private sendThiefReportedProperties(): [Promise<void>, () => void] {
```

Comment:
>sendThiefReportedProperties [](start = 10, length = 27)

I keep hitting the same set of questions on all of these "top level" promises.  `sendThiefReportedProperties` implies a single send, but it looks like this is a recurring operation.  I don't know the answer, but what's the right way to differentiate between doing something once, and doing it over-and-over again?
3 ideas:
1. function-level documentaiton.
2. naming conventions.
3. is there a way to be more explicit using typings?  e.g. make a new type for this with a name like `RecurringCancellableTestOperation` and the function would be defined as ` private sendThiefReportedProperties(): RecurringCancellableTestOperation {`
(OK, that name is terrible and this might be _less_ readable now that I read it, but there might be an idea here)